File: LeetCode/SlidingWindow/JavaScript/substringWithConcatenation.js
// You are given a string s and an array of strings words. All the strings of words are of the same length.

// A concatenated string is a string that exactly contains all the strings of any permutation of words concatenated.

// For example, if words = ["ab","cd","ef"], then "abcdef", "abefcd", "cdabef", "cdefab", "efabcd", and "efcdab" are all concatenated strings. "acdbef" is not a concatenated string because it is not the concatenation of any permutation of words.
// Return an array of the starting indices of all the concatenated substrings in s. You can return the answer in any order.

// Example 1:

// Input: s = "barfoothefoobarman", words = ["foo","bar"]

// Output: [0,9]

// Explanation:

// The substring starting at 0 is "barfoo". It is the concatenation of ["bar","foo"] which is a permutation of words.
// The substring starting at 9 is "foobar". It is the concatenation of ["foo","bar"] which is a permutation of words.

// Example 2:

// Input: s = "wordgoodgoodgoodbestword", words = ["word","good","best","word"]

// Output: []

// Explanation:

// There is no concatenated substring.

// Example 3:

// Input: s = "barfoofoobarthefoobarman", words = ["bar","foo","the"]

// Output: [6,9,12]

// Explanation:

// The substring starting at 6 is "foobarthe". It is the concatenation of ["foo","bar","the"].
// The substring starting at 9 is "barthefoo". It is the concatenation of ["bar","the","foo"].
// The substring starting at 12 is "thefoobar". It is the concatenation of ["the","foo","bar"].

// Constraints:

// 1 <= s.length <= 104
// 1 <= words.length <= 5000
// 1 <= words[i].length <= 30
// s and words[i] consist of lowercase English letters.

// TC:
// SC:

/**
 * @param {string} s
 * @param {string[]} words
 * @return {number[]}
 */
var findSubstring = function (s, words) {
  // Edge case
  if (!s.length || !words.length) return [];

  // Let's get a map of the words we need to check
  let word_freq = {};

  for (let word of words) {
    if (word_freq[word]) {
      word_freq[word] += 1;
    } else {
      word_freq[word] = 1;
    }
  }

  // We need these lengths in order to update our window
  const word_len = words[0].length; // represents the length of word at index 0
  const window_len = words.length * word_len; // represents the length of the window

  let res = [];

  // We'll loop through each char of s and slide the window
  // Let's create a window by using r + window_len
  // In this window, let's set the current word of s and check of each word is in our need word list
  for (let r = 0; r < s.length - window_len + 1; r++) {
    let substr_freq = {};
    let l = r;

    // let's check window length at a time
    while (l < r + window_len) {
      // in our window, let's check the word we need
      let cur_word = s.slice(l, l + word_len);

      // If the current word does not exist in our required word map, we break the while loop
      if (!(cur_word in word_freq)) {
        // console.log("Not here!")
        break;
      }

      // console.log(cur_word)

      // At this point, the current word is in our required word list
      // Now, let's create a map of our current word list.. This are the words we see in the window
      if (substr_freq[cur_word]) {
        substr_freq[cur_word] += 1;
      } else {
        substr_freq[cur_word] = 1;
      }

      // console.log(substr_freq)

      // Let's check if the number of words we see in our current word list is greater than the required word list
      // If so, we found all the words we need and let's break out of the while loop
      if (substr_freq[cur_word] > word_freq[cur_word]) {
        break;
      }

      // We keep going and slide our window
      l += word_len;
    }

    // At ths point, we exited the while loop and our l and r + window_len pointer overlapped
    // So we push index r to our results array
    if (l === r + window_len) {
      res.push(r);
    }
  }

  return res;
};

File: LeetCode/SlidingWindow/java/substringWithConcatenation.java
package LeetCode.SlidingWindow.java;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

record substringWithConcatenationRecord(String s, String[] words, List<Integer> expected) {
}

public class substringWithConcatenation {

    public List<Integer> findSubstring(String s, String[] words) {
        int input = 0;
        if (input == 0) {
            System.out.println("findSubstring1");
            return findSubstring1(s, words);
        } else {
            System.out.println("findSubstring2");
            return findSubstring2(s, words);
        }
    }

    public List<Integer> findSubstring1(String s, String[] words) {
        // edge case
        if (s.length() == 0 || words.length == 0) {
            return List.of();
        }

        // Let's grab a word freq map
        Map<String, Integer> word_freq = new HashMap<>();

        for (String word : words) {
            word_freq.put(word, word_freq.getOrDefault(word, 0) + 1);
        }
        // Map<String, Long> word_freq = Arrays.stream(words)
        // .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));

        // System.out.println("word_freq: " + word_freq.toString());

        // We need these lengths in order to update our window
        final int word_len = words[0].length(); // represents the length of word at index 0
        final int window_len = words.length * word_len; // represents the length of the window we need to check.

        List<Integer> res = new ArrayList<>();

        // We'll loop through each char of s and slide the window
        // Let's create a window by using r + window_len
        // In this window, let's set the current word of s and check of each word is in
        // our need word list
        for (int r = 0; r < s.length() - window_len + 1; r += 1) {
            Map<String, Integer> subStr_freq = new HashMap<>();
            int l = r;

            // let's check window length at a time
            while (l < r + window_len) {
                // in our window, let's check the word we need
                String cur_word = s.substring(l, l + word_len);

                // If the current word does not exist in our required word map, we break the
                // while loop
                if (!word_freq.containsKey(cur_word)) {
                    break;
                }

                // At this point, the current word is in our required word list
                // Now, let's create a map of our current word list.. This are the words we see
                // in the window
                // if (subStr_freq.containsKey(cur_word)) {
                // subStr_freq.put(cur_word, subStr_freq.get(cur_word) + 1);
                // } else {
                // subStr_freq.put(cur_word, 1);
                // }
                subStr_freq.put(cur_word, subStr_freq.getOrDefault(cur_word, 0) + 1);

                // Let's check if the number of words we see in our current word list is greater
                // than the required word list
                // If so, we found all the words we need and let's break out of the while loop
                if (subStr_freq.get(cur_word) > word_freq.get(cur_word)) {
                    break;
                }

                // We keep going and slide our window
                l += word_len;
            }

            // At ths point, we exited the while loop and our l and r + window_len pointer
            // overlapped
            // So we push index r to our results array
            if (l == r + window_len) {
                res.add(r);
            }
        }

        return res;
    }

    public List<Integer> findSubstring2(String s, String[] words) {
        List<Integer> result = new ArrayList<>();
        if (s.length() == 0 || words.length == 0)
            return result;

        int wordLen = words[0].length();
        int totalWords = words.length;
        // int totalLen = wordLen * totalWords;

        Map<String, Integer> wordCount = new HashMap<>();
        for (String word : words) {
            wordCount.put(word, wordCount.getOrDefault(word, 0) + 1);
        }

        for (int i = 0; i < wordLen; i++) {
            int left = i, right = i;
            int count = 0;
            Map<String, Integer> seen = new HashMap<>();

            while (right + wordLen <= s.length()) {
                String word = s.substring(right, right + wordLen);
                right += wordLen;

                if (wordCount.containsKey(word)) {
                    seen.put(word, seen.getOrDefault(word, 0) + 1);
                    count++;

                    while (seen.get(word) > wordCount.get(word)) {
                        String leftWord = s.substring(left, left + wordLen);
                        seen.put(leftWord, seen.get(leftWord) - 1);
                        count--;
                        left += wordLen;
                    }

                    if (count == totalWords) {
                        result.add(left);
                    }
                } else {
                    seen.clear();
                    count = 0;
                    left = right;
                }
            }
        }

        return result;
    }

    public static void main(String[] args) {
        substringWithConcatenationRecord input = new substringWithConcatenationRecord(
                "barfoothefoobarman",
                new String[] { "foo", "bar" },
                List.of(0, 9));
        substringWithConcatenation.testSolution(input);

        input = new substringWithConcatenationRecord(
                "wordgoodgoodgoodbestword",
                new String[] { "foo", "bar" },
                List.of());
        substringWithConcatenation.testSolution(input);

        input = new substringWithConcatenationRecord(
                "barfoofoobarthefoobarman",
                new String[] { "bar", "foo", "the" },
                List.of(6, 9, 12));
        substringWithConcatenation.testSolution(input);
    }

    private static void testSolution(substringWithConcatenationRecord input) {
        System.out.println("Input: s: " + input.s());
        System.out.println("words: " + Arrays.toString(input.words()));
        System.out.println("Expected: " + input.expected());
        List<Integer> val = new substringWithConcatenation().findSubstring(input.s(), input.words());
        System.out.println("Result: " + val);
        System.out.println((input.expected().equals(val) ? "PASS" : "FAIL"));
        System.out.println("-".repeat(50));
    }
}
